Load schemas in separate resource from State

oliverguenther opened the pull request.

oliverguenther committed:
Simplify updateImmediately link

oliverguenther committed:
Autoload the schemas from the schema states

oliverguenther committed:
Fix tests

oliverguenther committed:
Ensure new work packages are handled in states correctly

oliverguenther committed:
Avoid additional watcher on display fields

We can now simply wait for the cache service to return the WP + loaded
schema

ulferts commented on frontend/app/components/api/api-v3/hal-resources/work-package-resource.service.ts:
```diff
@@ -50,7 +52,7 @@ interface WorkPackageResourceEmbedded {
   relations: CollectionResourceInterface;
   responsible: HalResource|any;
-  schema: HalResource|any;
+  schema: SchemaResource|any;
```
Is it possible to remove the `|any` declaration?

ulferts commented on frontend/app/components/api/api-v3/hal-resources/work-package-resource.service.ts:
```diff
@@ -146,6 +149,10 @@ export class WorkPackageResource extends HalResource {
 
+  /**
+   * Returns a reference to the
+   */
```
Something is missing here

ulferts commented on frontend/app/components/api/api-v3/hal-resources/work-package-resource.service.ts:
```diff
@@ -319,14 +310,12 @@ export class WorkPackageResource extends HalResource {
         this.assignNewValues(form.$embedded.payload);
 
+        this.schema = form.$embedded.schema;
+
```
Why this change?

ulferts commented on frontend/app/components/work-packages/work-package-cache.service.ts:
```diff
@@ -74,8 +83,10 @@ export class WorkPackageCacheService {
+    if (workPackageId.toString() !== 'new') {
+    state.putFromPromiseIfPristine(() =>
+        this.apiWorkPackages.loadWorkPackageById(workPackageId, forceUpdate));
+    }
```
The responsibilities between `WorkPackageCacheService` and `ApiWorkPackageService` are not well separated. For `list`, the schemas are loaded by the `WorkPackageCache` while the `ApiWorkPackageService` loads the schema if `loadWorkPackage` is called: https://github.com/opf/openproject/pull/5063/files#diff-030dc6d5af01418720e8f296bf3b048aR65

ulferts commented on frontend/app/components/api/api-v3/hal-resources/hal-resource.service.ts:
```diff
@@ -100,7 +101,27 @@ export class HalResource {
+   * Return the associated state to this HAL resource, if any.
+   */
+  public get state():State<HalResource>|null {
+    return null;
```
I like the line of thought of each object having a reference to it's state and it looks nice in the `SchemaResource`. But it is not used in the WorkPackageResource, I believe because of the `WorkPackageCacheService` which is inconsistent IMO.

oliverguenther replied:
I agree. For sake of consistency, I have added the state to the work package resource, even though it is not in use by the work package. We cannot use that particular method inside  `#loadWorkPackage` anyway.

My next goal would be to remove the WP cache service. We can instead create a compound state of WorkPackage and Schema that is updated whenever each state is updated. That would be a rather big change due to the numerous references to the cache service.

oliverguenther replied:
I had to remove the state for work packages again, since there is no process that loads the work packages if e.g., a child is added that is not yet loaded in the state. We have to refactor the cache service before re-introducing it.

ulferts reviewed (changes requested):
The change works and I like that it was possible to remove that much code.

What irks me though, maybe because it's been a while since I took a look at the code, is the mix of responsibilities between a Resource, its State and the service responsible for loading.

oliverguenther committed:
Review remarks

oliverguenther commented:
I've addressed most your remarks. I won't be able to resolve the increased responsibility of the cache service now without refactoring all other usages to it.
